### PR TITLE
`User management`: Add filter for registration number

### PR DESCRIPTION
**Pull request by dangcosta**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

This change allows the admin user to filter (besides the existing ones) by registration number and to sync the user info with the LDAP server.

### Description

I added a new filter option to the filter part of the user management and a new button that allows the admin to sync with the the LDAP server. 

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Log in to Artemis
2. Navigate to User Management
3. On the main page click on the LDAP button for users with registration number (it should be available only for the users with the LDAP profile active. It is not possible to test on the tests servers since they do not have a LDAP connection).
4. Open the filter option and filter users with registration number, without registration number and with none filter.

#### Code Review
- [x] Review 1
- [x] Review 2

#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|

### Screenshots

image
image

### CODE REVIEW

**Review by jschmautzer — DISMISSED** [review]
tested on ts2, worked as described

**Review by Marvin-Vitcu — APPROVED** [review]
Tested on ts1: filter works as intended

**Comment by Dominik-Weinzierl** [comment]
We should first fix the tests and finish the PR before reviewing it :+1:

File: src/main/webapp/i18n/de/user-management.json
Comment (Dominik-Weinzierl) [inline src/main/webapp/i18n/de/user-management.json:36]: I suggest to use a similar translation as we use in the other filters: `Filter die Nutzer:innen anhand ihrer Registrierungsnummer`

**Review by Dominik-Weinzierl — APPROVED** [review]
Only a minor translation suggestion - the rest looks good :+1: (I have not tested manually)

File: src/main/webapp/i18n/de/user-management.json
Comment (dangcosta) [reply src/main/webapp/i18n/de/user-management.json]: Thanks Dominik! I changed the translation accordingly.

File: src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java:205]:
```suggestion
    @PutMapping("users/{login:" + Constants.LOGIN_REGEX + "}/sync-ldap")
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java:207]:
```suggestion
    public ResponseEntity<UserDTO> syncUserViaLdap(@PathVariable String login) {
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java:206]: please add the profile LDAP, so that this method is only available with an active LDAP profile

File: src/main/webapp/app/core/user/user.service.ts
Comment (krusche) [inline src/main/webapp/app/core/user/user.service.ts:39]: remove ts-ignore

File: src/main/webapp/app/admin/user-management/user-management.component.ts
Comment (krusche) [inline src/main/webapp/app/admin/user-management/user-management.component.ts:573]:
```suggestion
    syncLdap(login: string) {
```

File: src/main/webapp/app/admin/user-management/user-management.component.html
Comment (krusche) [inline src/main/webapp/app/admin/user-management/user-management.component.html:179]: This button should only be available when the profile LDAP is active.
Please check this in the ngOnInit method of the corresponding ts file
(there are examples how to load this from profile service)

File: src/main/webapp/app/admin/user-management/user-management.component.html
Comment (krusche) [inline src/main/webapp/app/admin/user-management/user-management.component.html:180]: I suggest to use translations here and call it `Sync from LDAP` in English

File: src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java
Comment (dangcosta) [reply src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java:205]: Change accepted.

File: src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java
Comment (dangcosta) [reply src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java:207]: Change accepted.

File: src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java
Comment (dangcosta) [reply src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java:206]: Profile added.

File: src/main/webapp/app/core/user/user.service.ts
Comment (dangcosta) [reply src/main/webapp/app/core/user/user.service.ts]: Change accepted. When I remove the //@ts-ignore I got the following error message: TS2322: Type 'Observable<User>' is not assignable to type 'Observable<HttpResponse<User>>'.   Type 'User' is missing the following properties from type 'HttpResponse<User>': body, type, clone, headers, and 4 more. So I removed the HttpResponse and made only Observable<User>.

File: src/main/webapp/app/admin/user-management/user-management.component.ts
Comment (dangcosta) [reply src/main/webapp/app/admin/user-management/user-management.component.ts:580]: Change accepted

File: src/main/webapp/app/admin/user-management/user-management.component.html
Comment (dangcosta) [reply src/main/webapp/app/admin/user-management/user-management.component.html]: Changes adapted. One quick question, though. When you say available, you mean that the button should only appear to the profiles with LDAP active (*ngIf) or the button should always appear but for those who are not active it should be disabled (like the delete button for the current user)?

File: src/main/webapp/app/admin/user-management/user-management.component.html
Comment (dangcosta) [reply src/main/webapp/app/admin/user-management/user-management.component.html]: Changes accepted. In German, I used the translation "von LDAP synchronisieren"

File: src/main/webapp/app/admin/user-management/user-management.component.html
Comment (Dominik-Weinzierl) [inline src/main/webapp/app/admin/user-management/user-management.component.html:179]: I suggest, as Stephan mentioned, that the button should not be displayed, not just disabled.

**Review by jschmautzer — DISMISSED** [review]
tested on ts2 (everything except the LDAP button, since the testservers are not configured with a LDAP server), filter works as expected

File: src/main/webapp/app/admin/user-management/user-management.component.html
Comment (dangcosta) [reply src/main/webapp/app/admin/user-management/user-management.component.html]: @Dominik-Weinzierl , I changed the button accordingly

**Review by jschmautzer — DISMISSED** [review]
tested on ts2, LDAP button no longer there :) and filter still working as expected

**Review by Dominik-Weinzierl — APPROVED** [review]
Did NOT perform any testings but the code looks good to me :+1:

**Review by Marvin-Vitcu — APPROVED** [review]
Tested on TS1. Filter works as intended.
LDAP button is missing, but this is okay (as mentioned by another reviewer)

**Review by nairaarnez — APPROVED** [review]
Tested in TS1, works as described.

File: src/main/java/de/tum/in/www1/artemis/service/user/UserCreationService.java
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/user/UserCreationService.java:256]: This comment is wrong. The password is not synchronized to Artemis. Please change it

File: src/main/java/de/tum/in/www1/artemis/service/user/UserService.java
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/user/UserService.java:386]: This method is mostly duplicated from `createUserFromLdap`
I don't think this is a good idea. I suggest we combine both into one method.
Please notice that `createUserFromLdap` already handles the case that the user exists in the Artemis database and then updates the user accordingly

File: src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java:206]: I suggest to use there `userId` here instead of the login with the regex

File: src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java:212]: By using the `userId` above, you can use the method `findByIdWithGroupsAndAuthoritiesElseThrow` here which simplifies the code

File: src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java:215]: I don't think this is the intended functionality. What happens in the user's matriculation number is null? This would not work.

Instead I suggest to make this much simpler, have a look how JiraAuthenticationProvider:168 does that. I would propose the following code.

```
ldapUserService.ifPresent(service -> service.loadUserDetailsFromLdap(user));
userCreationService.saveUser(user)
```

**Review by krusche — CHANGES_REQUESTED** [review]
See my inline comments. We can definitely simplify the server code a lot

File: src/main/java/de/tum/in/www1/artemis/service/user/UserService.java
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/user/UserService.java:386]: When following the changes below, this method might not even be needed and could be removed

File: src/main/java/de/tum/in/www1/artemis/service/user/UserCreationService.java
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/user/UserCreationService.java:268]: When following the changes below, this method might not even be needed and could be removed

**Review by Dominik-Weinzierl — CHANGES_REQUESTED** [review]
I agree with Stephan, we should avoid code duplication.

**Comment by Dominik-Weinzierl** [comment]
There are still 4 Codacy issues, please fix them :+1:

**Review by Marvin-Vitcu — APPROVED** [review]
Code review. LGTM

**Review by nairaarnez — DISMISSED** [review]
LGTM:+1: Approved

**Review by jschmautzer — DISMISSED** [review]
re-approve (tested on ts2 -> filter still working, LDAP button not visible)

File: src/main/webapp/app/admin/user-management/user-management.component.ts
Comment (JohannesStoehr) [inline src/main/webapp/app/admin/user-management/user-management.component.ts:214]:
```suggestion
            this.isLdapProfileActive = profileInfo.activeProfiles && profileInfo.activeProfiles?.includes('ldap');
```

File: src/main/webapp/app/admin/user-management/user-management.component.ts
Comment (JohannesStoehr) [inline src/main/webapp/app/admin/user-management/user-management.component.ts:314]: What is value here? Is this the newFilter or something like that?
A better name or a mention in the comment above with @param would help

File: src/test/java/de/tum/in/www1/artemis/util/UserTestService.java
Comment (JohannesStoehr) [inline src/test/java/de/tum/in/www1/artemis/util/UserTestService.java:810]: I suggest using the Constant from UserRepository

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/admin/user-management/user-management.component.ts
Comment (dangcosta) [reply src/main/webapp/app/admin/user-management/user-management.component.ts]: Value changed to registrationNumber and @param included.

File: src/test/java/de/tum/in/www1/artemis/util/UserTestService.java
Comment (dangcosta) [reply src/test/java/de/tum/in/www1/artemis/util/UserTestService.java]: Adapted.

File: src/main/webapp/app/admin/user-management/user-management.component.ts
Comment (dangcosta) [reply src/main/webapp/app/admin/user-management/user-management.component.ts]: Adapted

**Review by JohannesStoehr — APPROVED** [review]
Thanks for incorporating my suggestions. Code looks good to me

**Review by JohannesStoehr — APPROVED** [review]
Code looks still good after merge and test fix

**Review by Marvin-Vitcu — APPROVED** [review]
Re-Approved.

**Review by Santia-go — APPROVED** [review]
Tested on TS2. 
It works as expected. 
Re approved!

File: src/main/java/de/tum/in/www1/artemis/repository/specs/UserSpecs.java
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/repository/specs/UserSpecs.java:113]: ❓Question: What happens in the case that I pass in both parameters as true?

File: src/main/java/de/tum/in/www1/artemis/repository/specs/UserSpecs.java
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/repository/specs/UserSpecs.java:119]: ❓Question: I am not an expert on the criteria builder but withoutRegistrationNumber would mean for me null` or empty (meaning '') and withRegistrationNumber would mean **Not**(null or empty).

`criteriaBuilder.notEqual(root.get(User_.REGISTRATION_NUMBER), "");` does not seem to handle the null case?

File: src/test/java/de/tum/in/www1/artemis/util/UserTestService.java
Comment (stefanwaldhauser) [inline src/test/java/de/tum/in/www1/artemis/util/UserTestService.java:966]: ❓Question: Just to confirm. You are also testing the case here where the registration number is null right?

**Review by stefanwaldhauser — APPROVED** [review]
Good Job! Some requested changes ❌, ideas 💡 and questions ❓. Only changes ❌ need to be addressed for approval.

File: src/main/java/de/tum/in/www1/artemis/repository/specs/UserSpecs.java
Comment (dangcosta) [reply src/main/java/de/tum/in/www1/artemis/repository/specs/UserSpecs.java:113]: It is not possible due to the client side implementation (you can toggle one of the options (which are with, without and none of the filters) but not pass both as true)

File: src/test/java/de/tum/in/www1/artemis/util/UserTestService.java
Comment (dangcosta) [reply src/test/java/de/tum/in/www1/artemis/util/UserTestService.java:966]: It worked for both cases

**Pull request merged**
